src/groups/GroupList.test.tsx

```diff
@@ -0,0 +1,186 @@
+import { DeepPartial } from '@terra-ui-packages/core-utils';
+import { fireEvent, waitFor } from '@testing-library/react';
+import { userEvent } from '@testing-library/user-event';
+import React from 'react';
+import { GroupList } from 'src/groups/GroupList';
+import { Ajax } from 'src/libs/ajax';
+import { CurrentUserGroupMembership } from 'src/libs/ajax/Groups';
+import { get as getStateHistory, update as updateStateHistory } from 'src/libs/state-history';
+import { asMockedFn, renderWithAppContexts as render } from 'src/testing/test-utils';
+
+jest.mock('src/libs/nav', (): typeof import('src/libs/nav') => ({
+  ...jest.requireActual('src/libs/nav'),
+  getLink: jest.fn((link) => link),
+}));
+
+type ErrorExports = typeof import('src/libs/error');
+const mockReportError = jest.fn();
+
+jest.mock(
+  'src/libs/error',
+  (): ErrorExports => ({
+    ...jest.requireActual('src/libs/error'),
+    reportError: (...args) => mockReportError(...args),
+  })
+);
+
+type AjaxContract = ReturnType<typeof Ajax>;
+jest.mock('src/libs/ajax');
+
+jest.mock('src/libs/state-history', (): typeof import('src/libs/state-history') => ({
+  ...jest.requireActual('src/libs/state-history'),
+  get: jest.fn().mockReturnValue({}),
+  update: jest.fn(),
+}));
+
+describe('GroupList', () => {
+  const memberGroup: CurrentUserGroupMembership = {
+    groupEmail: 'group1@email.com',
+    groupName: 'test-group1-name',
+    role: 'member',
+  };
+  const adminGroup: CurrentUserGroupMembership = {
+    groupEmail: 'group2@email.com',
+    groupName: 'test-group2-name',
+    role: 'admin',
+  };
+
+  it('renders the no groups message', async () => {
+    asMockedFn(Ajax).mockImplementation(
+      () =>
+        ({
+          Groups: {
+            list: jest.fn().mockResolvedValue([]),
+          },
+        } as DeepPartial<AjaxContract> as AjaxContract)
+    );
+    const { getByText } = render(<GroupList />);
+    await waitFor(() => expect(getByText('Create a group to share your workspaces with others.')).toBeDefined());
+  });
+
+  it('renders all groups in the list', async () => {
+    asMockedFn(Ajax).mockImplementation(
+      () =>
+        ({
+          Groups: {
+            list: jest.fn().mockResolvedValue([memberGroup, adminGroup]),
+          },
+        } as DeepPartial<AjaxContract> as AjaxContract)
+    );
+    const { getByText } = render(<GroupList />);
+    await waitFor(() => expect(getByText(memberGroup.groupName, { exact: false })).toBeDefined());
```

Comment:
Nit: in general, you can avoid the waitFors if you instead await the render.

Reply:
Or you may not even need to await the render…

Reply:
I don't know of any way to do that. The `render` function isn't async.
Otherwise ` const { getByText } = render(<GroupList />);` wouldn't work, because it would be assigning a promise to `getByText`.
Unless there's an alternative I'm not aware of?

I don't think that would work here anyway, because of the async data retrieval.

Reply:
We typically just use `screen.getByText` instead of getting it from the render function result.

And right, you don't await the render… I was remembering where sometimes I have to wrap the render in an `act` and await that. In general, you should just be able to call render (no await) and then immediately call `screen.getByText` without a `waitFor`.